x-pack/plugins/osquery/server/routes/action/create_action_route.ts

```diff
@@ -24,6 +31,63 @@ export const createActionRoute = (router: IRouter) => {
     },
     async (context, request, response) => {
       const esClient = context.core.elasticsearch.client.asInternalUser;
+      let selectedAgents: string[] = [];
+      const {
+        agentSelection: { allAgentsSelected, platformsSelected, policiesSelected, agents },
+      } = request.body as { agentSelection: AgentsSelection };
+      // TODO: fix up the types here
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const extractIds = ({ body }: Record<string, any>) =>
+        body.hits.hits
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          .map((o: Record<string, any>) => o._source.local_metadata?.elastic.agent.id)
+          .filter((e: string) => e);
+      if (allAgentsSelected) {
+        // make a query for all agent ids
+        const idRes = await esClient.search({
```

Comment:
I think it'd be safier to use `agentsService` provided by `fleet` plugin

Reply:
this turned out to be a hell of a lot clearer too, thanks for the pointer!